oximeter/db/src/model.rs

```diff
@@ -647,267 +665,412 @@ pub(crate) fn unroll_measurement_row(sample: &Sample) -> (String, String) {
     let timeseries_name = sample.timeseries_name.clone();
     let timeseries_key = crate::timeseries_key(sample);
     let measurement = &sample.measurement;
+    unroll_measurement_row_impl(timeseries_name, timeseries_key, measurement)
+}
+
+pub(crate) fn unroll_measurement_row_impl(
+    timeseries_name: String,
+    timeseries_key: TimeseriesKey,
+    measurement: &Measurement,
+) -> (String, String) {
     let timestamp = measurement.timestamp();
     let extract_start_time = |measurement: &Measurement| {
         measurement
             .start_time()
             .expect("Cumulative measurements must have a start time")
     };
-    match measurement.datum() {
-        Datum::Bool(inner) => {
+
+    let datum = measurement.datum();
+    match measurement.datum_type() {
+        DatumType::Bool => {
+            let datum = match datum {
+                Datum::Bool(inner) => Some(DbBool::from(*inner)),
+                Datum::Missing(_) => None,
+                _ => unreachable!(),
+            };
```

Comment:
I think I see why this is happening, but this pattern (match on both `datum_type` + `datum`) is kinda a bummer.

We're basically trying to deal with a mismatch between `measurement.datum()` and `measuremnt.datum_type()`, even though they're both derived from the same field.

Why can't we match on the `datum`, like we did before?

```rust
match measurement.datum() {
   Datum::Bool(inner) => {
      // Handle the non-missing boolean case, like we did before
   }
   Datum::I8(inner) => {
      // Handle the non-missing i8 case, like we did before
   }
   ...
   Datum::Missing(missing) => {
      // Handle all the missing cases at once, read `datum_type` here if we need it
   }
}
```

Reply:
> but this pattern (match on both datum_type + datum) is kinda a bummer.

It sure is. I experimented with a bunch of ways to represent a missing sample in a way that includes all the information we need to insert a NULL record into the database. For example, the `start_time` for a cumulative measurement is included in the `Cumulative<T>` type, which means we can't store an `Option<Cumulative<T>>` since we'll then be missing the start time.

> Why can't we match on the datum, like we did before?

I think that works just fine too. I was trying to keep the conversion to each row type, e.g., `BoolMeasurementRow`, in one place, by generating an `Option`al value for that row at once. I don't feel particularly strongly about it, since this is a huge-but-rote match anyway, but your approach does mean we have two places that we do that construction. It does remove the need for `unreachable!()` though, which is nice. I'll give it a whirl now, thanks for the suggestion!

Reply:
Appreciate the cleanup - I know this is a bit more verbose, but it's easier to parse for humans IMO